ludwig/benchmarking/reporting.py

```diff
@@ -1,35 +1,166 @@
+        memory_so_far[device] += nbytes
+        max_so_far[device] = max(max_so_far[device], memory_so_far[device])
+        average_so_far[device] = (memory_so_far[device] + (average_so_far[device] * count_so_far[device])) / (
+            count_so_far[device] + 1
+        )
+        count_so_far[device] += 1
+    for device in count_so_far:
+        run_usage_info[f"torch_average_{device}_memory_usage"] = average_so_far[device]
+        run_usage_info[f"torch_max_{device}_memory_usage"] = max_so_far[device]
+    return run_usage_info
+
+
+def get_torch_op_time(events: List[profiler_util.FunctionEvent], attr: str) -> Union[int, float]:
+    """Get time torch operators spent executing for a list of events.
+
+    :param events: list of events.
+    :param attr: a FunctionEvent attribute. Expecting one of "cpu_time_total", "cuda_time_total".
+    """
+    if attr not in ["cpu_time_total", "cuda_time_total"]:
+        return -1
+
+    total = 0
+    for e in events:
+        if LUDWIG_TAG not in e.trace_name:
+            total += getattr(e, attr)
         else:
-            raise ValueError("Tag unrecognized. Please choose 'train' or 'evaluate'.")
+            total += get_torch_op_time(e.cpu_children, attr)
+    return total
+
+
+def get_device_run_durations(
+    function_event: profiler_util.FunctionEvent, run_usage_info: Dict[str, Any]
+) -> Dict[str, Any]:
+    """Get CPU and CUDA run durations for an event.
+
+    :param function_event: a function event instance.
+    :param run_usage_info: usage info for one execution of the tagged code block.
+    """
+    run_usage_info["torch_cpu_time"] = get_torch_op_time(function_event.cpu_children, "cpu_time_total")
+    run_usage_info["torch_cuda_time"] = get_torch_op_time(function_event.cpu_children, "cuda_time_total")
+    return run_usage_info
+
+
+def get_resource_usage_report(
+    main_kineto_events: List[_KinetoEvent],
+    main_function_events: List[profiler_util.FunctionEvent],
+    memory_events: List[Any],
+    info: Dict[str, Any],
+) -> Dict[str, Any]:
+    """Get relevant information from Kineto events and function events exported by the profiler.
+
+    :param main_kineto_events: list of main Kineto events.
+    :param main_function_events: list of main function events.
+    :param memory_events: list of memory events.
+    :param info: dictionary used to record resource usage metrics.
+    """
+    mem_records_acc = profiler_util.MemRecordsAcc(memory_events)
+    main_kineto_events = sorted(
+        (evt for evt in main_kineto_events if LUDWIG_TAG in evt.name()), key=lambda x: x.correlation_id()
+    )
+    main_function_events = sorted((evt for evt in main_function_events if LUDWIG_TAG in evt.name), key=lambda x: x.id)
+    assert [evt.id for evt in main_function_events] == [evt.correlation_id() for evt in main_kineto_events]
+    assert [evt.name for evt in main_function_events] == [evt.name() for evt in main_kineto_events]
```

Comment:
These asserts would cause a crash in production if the conditions aren't met. Are these asserts essential?

Reply:
They aren't. Removing them